models/auth/auth.go

```diff
@@ -15,14 +15,103 @@
 package auth
 
 import (
-	"github.com/fatedier/frp/models/config"
+	"fmt"
+
 	"github.com/fatedier/frp/models/consts"
 	"github.com/fatedier/frp/models/msg"
+
+	"github.com/vaughan0/go-ini"
 )
 
-type BaseAuth struct {
-	authenticateHeartBeats   bool
-	authenticateNewWorkConns bool
+type baseConfig struct {
+	// AuthenticationMethod specifies what authentication method to use to
+	// authenticate frpc with frps. If "token" is specified - token will be
+	// read into login message. If "oidc" is specified - OIDC (Open ID Connect)
+	// token will be issued using OIDC settings. By default, this value is "token".
+	AuthenticationMethod string `json:"authentication_method"`
+	// AuthenticateHeartBeats specifies whether to include authentication token in
+	// heartbeats sent to frps. By default, this value is false.
+	AuthenticateHeartBeats bool `json:"authenticate_heartbeats"`
+	// AuthenticateNewWorkConns specifies whether to include authentication token in
+	// new work connections sent to frps. By default, this value is false.
+	AuthenticateNewWorkConns bool `json:"authenticate_new_work_conns"`
+}
+
+func getDefaultBaseConf() baseConfig {
+	return baseConfig{
+		AuthenticationMethod:     "token",
+		AuthenticateHeartBeats:   false,
+		AuthenticateNewWorkConns: false,
+	}
+}
+
+func unmarshalBaseConfFromIni(conf ini.File) baseConfig {
```

Comment:
I prefer not put parse functions here. There is no need to coupled them together since how to unmarshal config is not releated to this package.

We may support more configure file types future. It will be a bit complicated to modify code everywhere.

Reply:
So you prefer the .ini unmarshal specifically should happen back in client_common and server_common?

Reply:
Yes.

Reply:
For common things, I agree. But for example, each option that is specific to an auth provider should be processed by that auth provider. So auth providers should also have an interface to parse the configuration and they should implement it.

Reply:
I agree with mpeterson, I think it's cleaner that `common` isn't aware of the OIDC / token specific configurations. And since we're doing that, I don't mind that the auth configuration resides here, like Proxy has unmarshalFromIni

Reply:
Not really. `AuthClientConfig` includes all auth provider's options. What one auth provider need to implement is how to validate options it needs, not how to parse it from configuration file. 

For example, once we want to support json format, we parse the `ServerConfig` once, no need to call each module's parsing funtions.

```go
type ServerConfig struct {
    ...
    Auth auth.AuthClientConfig `json:"auth"`
    ...
}

var config ServerConfig
json.Unmarshal(buf, &config)

authSetter := auth.NewAuthSetter(config.Auth)
```

Reply:
@fatedier but that will make server_common aware of every new protocol that we add, with every small configuration within, and that's the only reference this configuration will have outside of the `auth` module.

In addition, IDK if we want to give AuthClientConfig a name (e.g. Auth) because that suggests that the structure isn't flat, when in reality the .ini is flat.

Reply:
Yeah, I think this doesn't scale up if new protocols will be added in the future. In a way by doing so packages are "leaking" implementation details.

Reply:
Yes, you are right. It's annoying that we can't map the struct to ini file in a simple way.

Keep it here for ini now. I will try to support toml or other formats later.